ksql-api/src/main/java/io/confluent/ksql/api/server/BufferedPublisher.java

```diff
@@ -0,0 +1,284 @@
+      complete = true;
+    }
+  }
+
+  /**
+   * Subscribe a subscriber to this publisher. The publisher will allow at most one subscriber.
+   *
+   * @param subscriber The subscriber
+   */
+  @Override
+  public void subscribe(final Subscriber<? super T> subscriber) {
+    Objects.requireNonNull(subscriber);
+    if (Vertx.currentContext() == ctx) {
+      doSubscribe(subscriber);
+    } else {
+      ctx.runOnContext(v -> doSubscribe(subscriber));
+    }
+  }
+
+  /**
+   * Hook to allow subclasses to inject errors etc.
+   * This will be called before onNext is called on the subscriber to deliver an element
+   *
+   * @return true if processing should continue
+   */
+  protected boolean beforeOnNext() {
+    return true;
+  }
+
+  protected void sendError(final Exception e) {
+    checkContext();
+    try {
+      subscriber.onError(e);
+      cancelled = true;
+    } catch (Throwable t) {
+      logError("Exceptions must not be thrown from onError", t);
+    }
+  }
+
+  private void sendComplete() {
+    try {
+      cancelled = true;
+      subscriber.onComplete();
+    } catch (Throwable t) {
+      logError("Exceptions must not be thrown from onComplete", t);
+    }
+  }
+
+  private void doSubscribe(final Subscriber<? super T> subscriber) {
+    this.subscriber = subscriber;
+    try {
+      subscriber.onSubscribe(new Sub());
+    } catch (final Throwable t) {
+      sendError(new IllegalStateException("Exceptions must not be thrown from onSubscribe", t));
+    }
+  }
+
+  private void checkContext() {
+    if (Vertx.currentContext() != ctx) {
+      throw new IllegalStateException("On wrong context");
+    }
+  }
+
+  private void doSend() {
+    int numSent = 0;
+    while (!cancelled && demand > 0 && !buffer.isEmpty()) {
+      if (numSent < SEND_MAX_BATCH_SIZE) {
+        final T val = buffer.poll();
+        doOnNext(val);
+        numSent++;
+      } else {
+        // Schedule another batch async
+        ctx.runOnContext(v -> doSend());
+        break;
+      }
+    }
+
+    if (buffer.isEmpty() && !cancelled) {
+      if (complete) {
+        sendComplete();
+      } else if (demand > 0 && drainHandler != null) {
+        final Runnable handler = drainHandler;
+        ctx.runOnContext(v -> handler.run());
+        drainHandler = null;
+      }
+    }
+  }
+
+  private void doOnNext(final T val) {
+    if (!beforeOnNext()) {
+      return;
+    }
+    try {
+      subscriber.onNext(val);
+    } catch (final Throwable t) {
+      logError("Exceptions must not be thrown from onNext", t);
+    }
+    // If demand == Long.MAX_VALUE this means "infinite demand"
+    if (demand != Long.MAX_VALUE) {
+      demand--;
+    }
+  }
+
+  private void logError(final String message, final Throwable t) {
+    log.error(message, t);
+    cancelled = true;
+  }
+
+  private void doRequest(final long n) {
+    if (n <= 0) {
+      sendError(new IllegalArgumentException("Amount requested must be > 0"));
+    } else if (demand + n < 1) {
```

Comment:
I'm confused by this - according to the `Subscriber` docs, `n` is a _strictly positive_ number. Looking at where demand is set, I don't see how this can be less than 1.

Reply:
integer (long) overflow. I will add a comment to clarify.